xds/internal/balancer/clusterresolver/balancerconfig/configbuilder_test.go

```diff
@@ -0,0 +1,867 @@
+ * limitations under the License.
+ *
+ */
+
+package balancerconfig
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc/attributes"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/roundrobin"
+	"google.golang.org/grpc/balancer/weightedroundrobin"
+	"google.golang.org/grpc/internal/hierarchy"
+	internalserviceconfig "google.golang.org/grpc/internal/serviceconfig"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/xds/internal"
+	"google.golang.org/grpc/xds/internal/balancer/clusterimpl"
+	"google.golang.org/grpc/xds/internal/balancer/lrs"
+	"google.golang.org/grpc/xds/internal/balancer/priority"
+	"google.golang.org/grpc/xds/internal/balancer/weightedtarget"
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+)
+
+const (
+	testClusterName     = "test-cluster-name"
+	testLRSServer       = "test-lrs-server"
+	testMaxRequests     = 314
+	testEDSServcie      = "test-eds-service-name"
+	testEDSServiceName  = "service-name-from-parent"
+	testDropCategory    = "test-drops"
+	testDropOverMillion = 1
+
+	localityCount      = 5
+	addressPerLocality = 2
+)
+
+var (
+	testLocalityIDs []internal.LocalityID
+	testAddressStrs [][]string
+	testEndpoints   [][]xdsclient.Endpoint
+
+	testLocalitiesP0, testLocalitiesP1 []xdsclient.Locality
+)
+
+func init() {
+	for i := 0; i < localityCount; i++ {
+		testLocalityIDs = append(testLocalityIDs, internal.LocalityID{Zone: fmt.Sprintf("test-zone-%d", i)})
+		var (
+			addrs []string
+			ends  []xdsclient.Endpoint
+		)
+		for j := 0; j < addressPerLocality; j++ {
+			addr := fmt.Sprintf("addr-%d-%d", i, j)
+			addrs = append(addrs, addr)
+			ends = append(ends, xdsclient.Endpoint{
+				Address:      addr,
+				HealthStatus: xdsclient.EndpointHealthStatusHealthy,
+			})
+		}
+		testAddressStrs = append(testAddressStrs, addrs)
+		testEndpoints = append(testEndpoints, ends)
+	}
+
+	testLocalitiesP0 = []xdsclient.Locality{
+		{
+			Endpoints: testEndpoints[0],
+			ID:        testLocalityIDs[0],
+			Weight:    20,
+			Priority:  0,
+		},
+		{
+			Endpoints: testEndpoints[1],
+			ID:        testLocalityIDs[1],
+			Weight:    80,
+			Priority:  0,
+		},
+	}
+	testLocalitiesP1 = []xdsclient.Locality{
+		{
+			Endpoints: testEndpoints[2],
+			ID:        testLocalityIDs[2],
+			Weight:    20,
+			Priority:  1,
+		},
+		{
+			Endpoints: testEndpoints[3],
+			ID:        testLocalityIDs[3],
+			Weight:    80,
+			Priority:  1,
+		},
+	}
+}
+
+// TestBuildPriorityConfigMarshalled is a sanity check that the built balancer
+// config can be parsed. The behavior test is covered by
+// TestBuildPriorityConfig.
+func TestBuildPriorityConfigMarshalled(t *testing.T) {
+	const (
+		testClusterName     = "cluster-name-for-watch"
+		testEDSServiceName  = "service-name-from-parent"
+		testLRSServer       = "lrs-addr-from-config"
+		testMaxReq          = 314
+		testDropCategory    = "test-drops"
+		testDropOverMillion = 1
+	)
+
```

Comment:
A bunch of these are also defined as global. Could you please clean them up.

Reply:
Removed